### ISSUE TITLE
Consider introducing runtime backwards compatibility

### ISSUE THREAD

**Issue by Pimm** [mollie/mollie-api-node#319]
We are currently introducing [breaking](//github.com/mollie/mollie-api-node/pull/314) [changes](//github.com/mollie/mollie-api-node/pull/315) for the 4.0.0 release, and there are more to come.

These breaking changes:
 * reduce cause for confusion,
 * improve discoverability,
 * improve consistency with the rest of the JavaScript ecosystem, and
 * improve consistency with the PHP version of this client.

These breaking changes will cause compile-time errors for TypeScript users (good, as it facilitates the migration); and runtime errors for JavaScript users (frustrating).

We could consider implementing runtime backwards compatibility features ‒ that is, features which don't affect the exported type declarations. TypeScript users would still get compile-time errors; but JavaScript users _might_ just get a working version without migration.

Obviously, backwards compatibility cannot exist forever. It would be a temporary measure.

**Comment by janpaepke** [comment]
The danger you are referring to here is, that in contrast to TS changes, JS changes might go unnoticed when upgrading.
Adding runtime backwards compatibility (which should produce console warnings) might be a good intermediate step.

However if there are features, where this is not possible (for example if it conflicts with TS changes) I vote strongly against this. The reason being that it gives the user a false sense of security.

An obvious step is to prepare a comprehensive migration guide.
If we expect many users to be affected, we _might_ also consider preparing migration scripts.

In short, if we add backwards compatibility features, it should cover everything. If it doesn't, we should make a clean break.

My gut is leaning towards the latter - this is what major versions are for and we can expect devs to understand that.

**Comment by Pimm** [comment]
> The danger you are referring to here is, that in contrast to TS changes, JS changes might go unnoticed when upgrading.

Exactly. My main concern is that ‒ because we are renaming a lot of things but not _everything_ ‒ a quick and careless test from a user (developer) may give the false impression that their code works with 4.×.× without any changes.

We always write [a comprehensive migration guide](//github.com/mollie/mollie-api-node/blob/master/MIGRATION.md).

> In short, if we add backwards compatibility features, it should cover everything. If it doesn't, we should make a clean break.

Introducing a `customers_payments` → `customerPayments` runtime backwards compatibility hack is trivial. A similar hack for `withParent` would not be. I want to finally get rid of `withParent`: it was poor API design to begin with, and removing it will clean up the codebase significantly. `withParent` runtime backwards compatibility would deprive us of that cleanup.

**Comment by Pimm** [comment]
> As a consumer of a library, I highly prefer backwards compatible updates over "source code cleanup". Frankly, I don't care _that much_ about the state of the source code of a library, assuming the library works well and doesn't have an extremely large footprint.

That's a fair point. I would like to reiterate that the removals primarily serve to:
* reduce cause for confusion,
* improve discoverability,
* improve consistency with the rest of the JavaScript ecosystem, and
* improve consistency with the PHP version of this client.

As a bonus, the source cleans up because some of the removals (such as `withParent`) would make existing clutter in the codebase unnecessary. If we provide runtime backwards compatibility, then we lose out on that specific bonus.

To provide some context:

Anything scheduled to be removed, has been deprecated for at least a year already. `withParent` specifically has been deprecated for almost five years. They have, however, been deprecated through JSDoc instead of `console.warn` as suggested by @janpaepke. Personally, I am not so sure how appropriate the use of `console.warn` would be. Those messages likely end up in a log somewhere, and in my experience those logs tend to go unchecked as long as the service works as intended.

**Comment by Pimm** [comment]
I did some digging. Almost everything I want to remove was deprecated 1½+ years ago, including:
 * `orders_payments` (the alias for `orderPayments`) and similar :snake:-aliases,
 * `payments.delete` (the alias for `payments.cancel`) and similar endpoint-aliases,
 * `list.count` (the alias for JavaScript-native `list.length`), and
 * `payment.isOpen()` (the alias for `payment.status == PaymentStatus.open`) and similar unhelpful helpers.

There is one exception, which was deprecated 4½+ years ago:
* `withParent`

Knowing this, I think a reasonable approach is introducing runtime-only backwards compatibility features for everything _except_ `withParent`. This will force TypeScript users to migrate while kicking the can for JavaScript users ‒ unless those JavaScript users happen to be still using `withParent`.

How do we feel about this? And should we log to `console.warn`?

**Comment by edorivai** [comment]
This makes sense to me. I agree that JSDoc deprecations are enough. Once you introduce `console.warn`, you'll also inevitable receive requests asking "how to suppress these warnings?" :sweat_smile: 

Given that all these aliases have been deprecated for over a year, I think it's a good start to remove them from TS. JS will follow. Users that don't want to deal with the migration will be stuck on an older version of the library, which is also kinda how things go in general, makes sense to me.
